---
title: "PacBio 16S Amplicon sequencing analysis"
author: Aiswarya Prasad
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  prettydoc::html_pretty:
    theme: cayman
    <!-- other fun themes: architect leonids hpstr cayman -->
    highlight: github
    math: katex
    number_sections: true
    df_print: paged
    cold-folding: hide
    toc: true
---

# Setup

The qPCR data can be found on nas recherche - make sure to copy relavant data or files if nas recherche is not mounted. Also ensure that the prefix directory in the qPCR parsing script corresponds to the appropriate path in the machine on which this is being run.

```{r setup}
library(tidyverse)
library(ggplot2)
library(readxl)
library(viridis)
library(circlize)
library(hrbrthemes)
library(ggthemes)
library(RColorBrewer)
library(scales)
library(dplyr)
library(vegan)
library(ape)
library(ComplexHeatmap)
library(ggnewscale)
library(ggsignif)
library(plotly)
library(htmlwidgets)
library(dada2)
library(phyloseq)
library(DECIPHER)

# prefix_dir <- "/home/aiswarya/mnt/nas_recherche/" # for lab workstation
# prefix_dir <- "/nas/FAC/FBM/DMF/pengel/" # for curnagl
# working_dir <- paste0(prefix_dir, "spirit/D2c/aprasad/20220921_aprasad_PriorityEffectsExperimentPilot/03_PilotExperiment/07_PacBIO_sequencing")
working_dir <- paste0("/scratch/aprasad/2023-pacbio_sequence_analysis")
setwd(working_dir)
source("/nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/20220921_aprasad_PriorityEffectsExperimentPilot/03_PilotExperiment/05_qPCRs/qPCR_parse.R") # set path to directory as needed within the scripts
source("scripts/utilities.R")

metadata_path <- "/nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/20220921_aprasad_PriorityEffectsExperimentPilot/03_PilotExperiment/06_PCR_PacBio/PCR_plan.xlsx"

df_meta_pilot <- read_excel(metadata_path, sheet = 1) %>%
                  mutate(Treatment = ifelse(Sample == "Day14 - blank", "0-3", Treatment))
# write.csv(df_meta, file = "230313_metadata.csv", append = F, quote = F, row.names = F)
treatment_order <- c("AB-0", "A-0", "A-3", "A-3_14", "B-0", "B-3", "B-3_14", "AB-1", "BA-1", "AB-3", "BA-3", "0-3", "alq-A", "alq-AB", "alq-B", "alq-SW")
treatment_order_color <- list("AB-0" = "#6a3d9a", 
                        "A-0" = "#1f78b4", "A-3" = "#ff7f00", "A-3_14" = "#ff9f00",
                        "B-0" = "#a6cee3", "B-3" = "#fdbf6f", "B-3_14" = "#ffdf6f",
                        "AB-1" = "#33a02c", "BA-1" = "#b2df8a",
                        "AB-3" = "#e31a1c", "BA-3" = "#fb9a99",
                        "0-3" = "#cab2d6",
                        "alq-A" = "#ff7f00", "alq-AB" = "#6a3d9a", "alq-B" = "#fdbf6f", 
                        "alq-SW" = "#cab2d6"
                        )
# arranged such that they are in treatment order
ID_order <- c("2-1", "2-13", "3-1", "3-13", "1-1",
              "1-4", "2-4", "2-16", "3-4", "3-16",
              "1-8", "2-8", "2-20", "3-8", "3-20",
              "1-11", "1-23", "2-11", "2-23", "3-11",
              "1-5", "2-5", "2-17", "3-5", "3-17", "1-9",
              "2-9", "2-21", "3-9", "1-12", "1-24", "2-12",
              "2-24", "3-12", "1-2", "2-2", "2-14",
              "3-2", "3-14", "1-3", "2-3", "2-15", "3-3",
              "3-15", "1-6", "2-6", "2-18", "3-6",
              "3-18", "1-7", "2-7", "2-19", "3-7", "3-19",
              "1-10", "2-10", "2-22", "3-10", "3-21",
              "3-22", "1-13", "1-14", "1-15", "1-19",
              "1-16","1-17", "1-18", "1-20", "1-21", "1-22"
              )
system("mkdir -p 03_Dada2Preprocessed")
system("mkdir -p 04_Results")
system("mkdir -p 04_Results/Figures")
system("mkdir -p 04_Results/RDS")
system("mkdir -p 04_Results/Taxonomy")
```

# Introduction

PCR amplicons were pooled as described in other directories in the project directory (`/nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/20220921_aprasad_PriorityEffectsExperimentPilot/03_PilotExperiment/`). Sequencing was done at the GTF, PacBio CCS. Below is a summary of the qPCR values.

## qPCR analysis summary

```{r qPCR_values_plot}
ggplot() +
  geom_point(data = combined_df,
             aes(y = factor(ID, ID_order),
                 x = bac_copies_norm,
                 color = factor(Treatment, treatment_order),
                 shape = Sample_type
         ), size = 3, stat = "identity"
         ) +
  labs(y = "Sample ID", x = "Bacterial copynumber (Normalized)", color = "Treatment") +
  scale_color_manual(values=treatment_order_color) +
    scale_x_continuous(trans = "log10", labels = function(x) parse(text=paste("10^",round(log10(x), 2)))) +
    # scale_y_log10(minor_breaks = unique(as.numeric(1:10 %o% 10 ^ (0:3)))) +
    geom_hline(yintercept = LOD) +
    make_theme(setCol = F, setFill = F, palettecolor = "Pastel1",
               axis_x_title = 18,
               axis_y_title = 18, guide_nrow = 3,
               y_size = 10,
               x_size = 18)
ggsave("04_Results/Figures/qPCR_copy_numbers.pdf")
```

The samples were pooled in the library based on the intensity of the bands of the respective samples after PCR. The sequencing yields are plotted below.

```{r}
raw_reads_renamed_path <- "01_ReadsRenamed/"
raw_reads_renamed <- as.character(lapply(ID_order, function(x) get_reads_path(x)))
# read_counts_df <- data.frame("ID" = ID_order) %>%
#                     mutate(raw_reads_path = Vectorize(get_reads_path)(ID)) %>%
#                     mutate(raw_reads = Vectorize(get_number_of_reads)(raw_reads_path)) %>%
#                     select(!raw_reads_path)
# saveRDS(read_counts_df, "RDS/read_counts_df_raw.RDS")
read_counts_df <- readRDS("04_Results/RDS/read_counts_df_raw.RDS")
df_volume_reads <- df_meta_pilot %>%
                    filter(ID %in% ID_order) %>%
                      select(ID, Sample, Volume) %>%
                        left_join(read_counts_df)
ggplot() +
  geom_bar(data = df_volume_reads,
             aes(y = factor(ID, ID_order),
                 x = raw_reads,
                 fill = factor(Volume, c(15, 10, 5))
         ), stat = "identity"
         ) +
  labs(y = "Sample ID", x = "Number of reads", fill = "Volume added (uL)    ") +
    scale_x_continuous(labels=unit_format(unit = "K", scale = 1e-3)) +
    make_theme(setCol = F, setFill = T, palettecolor = "Pastel1",
               axis_x_title = 18,
               axis_y_title = 18, guide_nrow = 1,
               y_size = 10,
               x_size = 18)
ggsave("04_Results/Figures/Number_of_reads_by_volume.pdf")

ggplot() +
  geom_bar(data = df_volume_reads,
             aes(y = factor(ID, ID_order),
                 x = raw_reads,
                 fill = factor(Volume, c(15, 10, 5))
         ), stat = "identity"
         ) +
  labs(y = "Sample ID", x = "Number of reads", fill = "Volume added (uL)    ") +
    scale_x_continuous(labels=unit_format(unit = "K", scale = 1e-3)) +
    scale_y_discrete(labels= ID_order %>% as.data.frame() %>% left_join(df_volume_reads, by = c("." = "ID")) %>% pull(Sample)) +
    make_theme(setCol = F, setFill = T, palettecolor = "Pastel1",
               axis_x_title = 18, y_hj = 1,
               axis_y_title = 18, guide_nrow = 1,
               y_size = 10,
               x_size = 18)
ggsave("04_Results/Figures/Number_of_reads_by_volume_labelled_name.pdf")
```

# Preprocessing reads

## Remove primers and reorient reads

First, primers must be removed and files re-oriented. The sequences of the primers can be found https://www.pacb.com/wp-content/uploads/Procedure-checklist-Amplification-of-bacterial-full-length-16S-rRNA-gene-with-barcoded-primers.pdf 
These will have the phrase `_primers_removed` and will be placed in the folder `03_Dada2Preprocessed`
barcodes may have already been removed upon demultiplexing so use just the part of the primer after the barcode.

```{r}
primer_remove_in_files = as.character(lapply(ID_order, function(x) paste0("01_ReadsRenamed/", x, "_reads.fastq.gz") ))
primer_remove_out_files = as.character(lapply(ID_order, function(x) paste0("03_Dada2Preprocessed/", x, "_primers_removed.fastq.gz")))
f_primer <- "AGRGTTYGATYMTGGCTCAG"
r_primer <- "RGYTACCTTGTTACGACTT"
r_primer_comp <- rc(r_primer)
# primer_removal_summary <- removePrimers(primer_remove_in_files, primer_remove_out_files,
#               primer.fwd = f_primer,
#               primer.rev = r_primer_comp,
#               verbose = TRUE)
# saveRDS(primer_removal_summary, "RDS/primer_removal_summary.rds")
primer_removal_summary <- readRDS("04_Results/RDS/primer_removal_summary.rds")
```

## Filter reads based on length

### Read length distribution before

The adapters but not primers appear to have been removed upon CCS analysis and demultiplexing.
The step to get read lengths takes a long time so save the output and reload for next time.

```{r}
# read_lengths <- lapply(raw_reads_renamed, function(x) nchar(getSequences(x)))
# saveRDS(read_lengths, "RDS/read_lengths.rds")
read_lengths <- readRDS("04_Results/RDS/read_lengths.rds")
pdf("04_Results/Figures/Read_length_histogram.pdf", width = 11, height = 8)
hist(do.call(c, read_lengths), 100)
dev.off()
hist(do.call(c, read_lengths), 100)
```

### Filtering by length

We expect and amplicon size in the range of 1500. So make a cut-off of 1000 minimum and 1600 maximum.
This has already bee performed and will not be repeated when knitting.

```{r}
filt_read_files <- as.character(lapply(ID_order, function(x) paste0("03_Dada2Preprocessed/", x, "_filt_reads.fastq.gz")))
# trimming_summary <- filterAndTrim(primer_remove_out_files, filt_read_files, minLen=1000, maxLen=1600, rm.phix=FALSE, multithread=T)
# saveRDS(trimming_summary, "RDS/trimming_summary.rds")
trimming_summary <- readRDS("04_Results/RDS/trimming_summary.rds")
```

### Read length distribution after

```{r}
# read_lengths_filt <- lapply(filt_read_files, function(x) nchar(getSequences(x)))
# saveRDS(read_lengths_filt, "RDS/read_lengths_filt.rds")
read_lengths_filt <- readRDS("04_Results/RDS/read_lengths_filt.rds")
pdf("04_Results/Figures/Read_length_filtered_histogram.pdf", width = 11, height = 8)
hist(do.call(c, read_lengths_filt), 100)
dev.off()
hist(do.call(c, read_lengths_filt), 100)
```

## Dada2 quality plots

```{r}
pdf("04_Results/Figures/QualityProfilePlot1-5.pdf", width = 11, height = 8)
plotQualityProfile(filt_read_files[1:5])
dev.off()
plotQualityProfile(filt_read_files[1:5])
```

## Dereplicate and learn errors

```{r}
# derepd <- derepFastq(filt_read_files, verbose = T)
# errors <- learnErrors(derepd, errorEstimationFunction=PacBioErrfun, BAND_SIZE=32, multithread=TRUE)
# saveRDS(errors, "RDS/errors.rds")
errors <- readRDS("04_Results/RDS/errors.rds")
plotErrors(errors)
# dada2 <- dada(derepd, err=errors, BAND_SIZE=32, multithread=T)
# saveRDS(dada2, "RDS/dada2.rds")
dada2 <- readRDS("04_Results/RDS/dada2.rds")
seqtable <- makeSequenceTable(dada2); dim(seqtable)
seqtable_nochim <- removeBimeraDenovo(seqtable, method = "consensus", multithread = TRUE, verbose = TRUE); dim(seqtable)
sum(seqtable_nochim)/sum(seqtable)
# hist(as.numeric(lapply(colnames(seqtable_nochim), function(x) nchar(x))))
```

## Summarise pre-processing

```{r}
# df_summary <- primer_removal_summary %>% as.data.frame() %>%
#                 mutate(ID = Vectorize(remove_extension)(rownames(.), "_reads.fastq.gz")) %>%
#                    pivot_longer(!ID, values_to = "reads", names_to = "type") %>%
#                     mutate(type = ifelse(type == "reads.in", "0_Raw.reads", "1_Primer.removed")) %>%
#                       left_join(df_meta_pilot, by = c("ID" = "ID"))
# df_summary_trimming <- trimming_summary %>% as.data.frame %>% 
#                                   mutate(ID = Vectorize(remove_extension)(rownames(.), "_primers_removed.fastq.gz")) %>%
#                                   pivot_longer(!ID, values_to = "reads", names_to = "type") %>%
#                                   mutate(type = ifelse(type == "reads.in", "Trimming.in", "2_Trimmed")) %>%
#                                   left_join(df_meta_pilot, by = c("ID" = "ID")) %>%
#                                   filter(type != "Trimming.in")
# dada_denoise_summary <- data.frame("V1" = sapply(dada2, function(x) sum(getUniques(x))) %>% as.matrix) %>%
#                           mutate("4_Denoised" = V1) %>% select(!V1) %>%
#                           cbind("5_No.chimera" = rowSums(seqtable_nochim)) %>%
#                           mutate(ID = Vectorize(remove_extension)(rownames(.), "_filt_reads.fastq.gz")) %>%
#                             pivot_longer(!ID, values_to = "reads", names_to = "type")
# df_summary <- bind_rows(rbind(df_summary, df_summary_trimming), dada_denoise_summary) %>%
#                 select(ID, type, reads) %>%
#                 left_join(df_meta_pilot %>% select(ID, Treatment)) %>%
#                 mutate(Treatment_type = ifelse(grepl("alq", Treatment), "aliquot", Treatment))
# saveRDS(df_summary, "04_Results/RDS/df_summary.rds")
df_summary <- readRDS("04_Results/RDS/df_summary.rds")

ggplot() +
  geom_bar(data = df_summary,
           aes(y = factor(ID, ID_order),
               x = reads,
               fill = type,
               group = type
           ), color = "black",
           stat = "identity", position = "dodge"
  ) +
  facet_wrap(~ Treatment, scale = "free_y") +
  scale_x_continuous(labels=unit_format(unit = "K", scale = 1e-3)) +
  labs(x = "Number of reads", y = "Sample", fill = "Preprocessing step    ") +
    make_theme(setFill = T, setCol = F, x_angle = 30)
ggsave("04_Results/Figures/summary_of_preprocessing.pdf")
```

```{r}
seqtable_out <- seqtable_nochim %>% as.data.frame %>%
            rownames_to_column("ID") %>%
            mutate(ID = Vectorize(remove_extension)(ID, "_filt_reads.fastq.gz")) %>%
              column_to_rownames("ID") %>%
              as.matrix()

hist(as.numeric(lapply(colnames(seqtable_out[, colSums(seqtable_out != 0) > 0]), function(x) nchar(x))), 100)
hist(as.numeric(lapply(colnames(seqtable_out[, colSums(seqtable_out != 0) > 1]), function(x) nchar(x))), 100)



df_strains <- read_excel("/nas/FAC/FBM/DMF/pengel/spirit/D2c/aprasad/20220921_aprasad_PriorityEffectsExperimentPilot/03_PilotExperiment/02_CalculateDilutionsOD.xlsx", sheet = 2)
genome_names <- c(df_strains$pair_a, df_strains$pair_b) %>% as.data.frame %>%
                  filter(startsWith(., "ESL")) %>%
                    pull
df_strains_info <- data.frame("Genome" = genome_names)
identical_copies <- c("ESL0824", "ESL0200", "ESL0197", "ESL0820", "ESL0170", "ESL0199", "ESL0819","ESL0185", "ESL0186")
non_identical_copies <- c("ESL0183", "ESL0184")
single_copies <- c("ESL0825", "ESL0822", "ESL0827", "ESL0295", "ESL0198", "ESL0394", "ESL0263", "ESL0262", "ESL0350", "ESL0261")
community_ab <- c("ESL0197")
community_a <- c("ESL0825","ESL0822","ESL0824","ESL0827","ESL0200","ESL0295","ESL0185","ESL0183","ESL0184","ESL0186")
community_b <- c("ESL0820", "ESL0170", "ESL0198", "ESL0199", "ESL0819", "ESL0394", "ESL0263", "ESL0262", "ESL0350", "ESL0261")
# database_sequences <- as.character(lapply(df_strains_info$Genome, function(x) system(paste0("ls ", working_dir, "/database/16S_sequences/*.fa | grep ", paste0(x, "-1")), intern = T)))
# x <- database_sequences[[12]]
# hist(as.numeric(
#       lapply(database_sequences,
#              function(x) system(paste0("cat ", x, " | grep -v '>' | wc -c"), intern = T)
#           )
#       )
#     )
```

```{r}
# system("wget https://zenodo.org/record/3986799/files/silva_nr99_v138_wSpecies_train_set.fa.gz  -O database/silva_nr99_v138_wSpecies_train_set.fa.gz")
silva_DB_assignTax <- "database/silva_nr99_v138_wSpecies_train_set.fa.gz"
# taxonomySilva <- assignTaxonomy(seqtable_nochim, silva_DB_assignTax, minBoot = 80, multithread=TRUE, verbose = TRUE)
# saveRDS(taxonomySilva, "RDS/taxonomySilva.rds")
taxonomySilva <- readRDS("04_Results/RDS/taxonomySilva.rds")
formatted_DB_assignSpec <- "/scratch/aprasad/2023-pacbio_sequence_analysis/database/pilot_experiment_sequences_for_species_assignment.fasta"
# taxonomyCommunity <- assignSpecies(taxonomySilva, formatted_DB_assignSpec, allowMultiple = T, verbose = T)
# saveRDS(taxonomyCommunity, "RDS/taxonomyCommunity.rds")
taxonomyCommunity <- readRDS("04_Results/RDS/taxonomyCommunity.rds")
taxonomy_df <- taxonomySilva %>% 
                as.data.frame %>%
                  rownames_to_column("ASV") %>%
                  left_join(taxonomyCommunity %>%
                              as.data.frame() %>%
                                mutate(SDP = Genus) %>%
                                mutate(Strain = Species) %>%
                                  select(!c(Genus, Species)) %>%
                                    rownames_to_column("ASV")
                                ) %>%
                                mutate(Community = Vectorize(get_community)(Strain))
write.csv(file="04_Results/Taxonomy/Taxonomy.csv", taxonomy_df)
```

```{r}
ranks_assigned_df <- data.frame(rank = c("Total", colnames(taxonomy_df))) %>%
                        mutate(num_ASVs = Vectorize(get_number_assigned)(rank))
ggplot() +
  geom_bar(data = ranks_assigned_df,
           aes(x = factor(rank, c("Total", colnames(taxonomy_df))),
               y = num_ASVs,
               fill = factor(rank, c("Total", colnames(taxonomy_df)))
           ), stat = "identity"
          ) +
    labs(x = "Rank", y = "Number of ASVs", fill = "Rank") +
      make_theme(max_colors = length(unique(ranks_assigned_df$rank)), 
                 x_angle = 30, x_hj = 1, x_vj = 1)
      ggsave("04_Results/Figures/Assigned_ranks.pdf")

ggplot() +
  geom_bar(data = taxonomy_df,
           aes(x = Phylum,
               fill = Phylum
           ), stat = "count"
          ) +
    labs(x = "Phylum", y = "Number of ASVs", fill = "Phylum") +
      make_theme(x_angle = 30)
      ggsave("04_Results/Figures/Phylum_summary.pdf")

ggplot() +
  geom_bar(data = taxonomy_df,
           aes(x = Family,
               fill = Family
           ), stat = "count"
          ) +
    labs(x = "Family", y = "Number of ASVs", fill = "Family") +
      make_theme(x_angle = 30, x_vj = 1, x_hj = 1, 
                 guide_nrow = 4,
                 x_size = 7,
                 leg_size = 6,
                 max_colors = length(unique(taxonomy_df$Family)))
      ggsave("04_Results/Figures/Family_summary.pdf")
```

```{r}
samples_df <- df_meta_pilot %>% 
                filter(!is.na(Sample)) %>%
                filter(!is.na(ID)) %>%
                column_to_rownames("ID")
taxonomy_mat <- taxonomy_df %>%
                  column_to_rownames("ASV") %>%
                    as.matrix
otu_mat <- seqtable_nochim %>% as.data.frame %>%
            rownames_to_column("ID") %>%
            mutate(ID = Vectorize(remove_extension)(ID, "_filt_reads.fastq.gz")) %>%
              column_to_rownames("ID") %>%
              as.matrix()
ps_raw <- phyloseq(otu_table(otu_mat, taxa_are_rows=F), 
               sample_data(samples_df), 
               tax_table(taxonomy_mat))

dna <- Biostrings::DNAStringSet(taxa_names(ps_raw))
names(dna) <- taxa_names(ps_raw)

ps_raw <- merge_phyloseq(ps_raw, dna)
taxa_names(ps_raw) <- paste0("ASV", seq(ntaxa(ps_raw)))
ps <- ps_raw

table = merge(tax_table(ps),t(otu_table(ps)), by="row.names")
write.table(table, "04_Results/Taxonomy/ASVtable.txt", sep="\t", row.names = F)

# Export to FASTA with Biostrings
Biostrings::writeXStringSet(refseq(ps), "04_Results/Taxonomy/phyloseq_ASVs.fasta",append=FALSE, format="fasta")
# save Phyloseq object
# saveRDS(ps, 'RDS/phyloseq_object_raw.rds')
# Check that you can import it
ps_raw <- readRDS("04_Results/RDS/phyloseq_object_raw.rds")
```

```{r}
ps_rel <- transform_sample_counts(ps, function(x) x / sum(x) )
abundant_taxa <- taxa_sums(ps_rel)[which(taxa_sums(ps_rel) > 0.005)]
ps_abundant_taxa <- prune_taxa(names(abundant_taxa), ps_rel)
# plot_bar(ps_abundant_taxa, fill = "Genus")
# plot_bar(ps_abundant_taxa, fill = "Strain")
# plot_bar(ps_abundant_taxa, fill = "Community    ")
```

Many ASVs appear to be from sequences that are unknown.
A lot of these "unknown" ASVs are not from contaminants but rather Lactobacillus or Bifidiobacterium.
These could simply be variants of sequences from sequences genomes which were lost out because of the assembly not distinctly assembling them. This is particularly likely in the case of genomes that are likely to have multiple copies.
We know from the aliquots being sequenced, which ASVs are from community A and which ASVs are from community B.

```{r}
adjusted_copy_numbers <- psmelt(ps) %>% 
    filter(Abundance > 0) %>%
    mutate(Species_name = paste0(Genus, " ", Species)) %>%
    left_join(combined_df %>% select(ID, bac_copies_norm), by = c("Sample" = "ID")) %>%
    left_join(df_summary %>% filter(grepl("5_", type)) %>% select(ID, reads), by = c("Sample" = "ID")) %>% 
    unique() %>%
    mutate(rel_ab = 100*Abundance/reads) %>%
      mutate(abs_ab = rel_ab*bac_copies_norm/min(bac_copies_norm)*as.numeric(Volume)) %>%
      arrange(Community, Genus, SDP, Strain)

adjusted_copy_numbers_aliquots <- adjusted_copy_numbers %>%
                                    filter(grepl("alq", Treatment)) %>%
                                      mutate(Community_inferred = Vectorize(infer_community_from_aliquot)(OTU))

abundance_mat_alq <- adjusted_copy_numbers_aliquots %>%
                      select(sample_Sample, OTU, abs_ab) %>%
                        unique %>%
                        pivot_wider(names_from = OTU, values_from = abs_ab, values_fn = as.numeric) %>%
                        replace(is.na(.), 0) %>%
                        column_to_rownames("sample_Sample") %>%
                        as.matrix

treatment_names <- rownames(abundance_mat_alq) %>% as.data.frame() %>% left_join(psmelt(ps) %>% ungroup() %>% select(sample_Sample, Treatment) %>% unique, by = c(. = "sample_Sample")) %>% pull(Treatment)
genus_names <- colnames(abundance_mat_alq) %>% as.data.frame() %>% left_join(psmelt(ps) %>% ungroup() %>% select(OTU, Genus) %>% unique, by = c(. = "OTU")) %>% pull(Genus)
species_names <- colnames(abundance_mat_alq) %>% as.data.frame() %>% left_join(psmelt(ps) %>% ungroup() %>% select(OTU, Species, Genus) %>% unique %>% mutate(Species_name = paste0(Genus, " ", Species)), by = c(. = "OTU")) %>% pull(Species_name)
SDP_names <- colnames(abundance_mat_alq) %>% as.data.frame() %>% left_join(psmelt(ps) %>% ungroup() %>% select(OTU, SDP) %>% unique, by = c(. = "OTU")) %>% pull(SDP)
strain_name <- colnames(abundance_mat_alq) %>% as.data.frame() %>% left_join(psmelt(ps) %>% ungroup() %>% select(OTU, SDP, Strain) %>% unique %>% mutate(Strain_name = paste0(SDP, "--", Strain)), by = c(. = "OTU")) %>% pull(Strain_name)
column_ba = HeatmapAnnotation(SDP = SDP_names,
                              Strain = strain_name,
                              Species = species_names,
                       col = list(SDP = SDPColors,
                                  Strain = strainColors,
                                  Species = extend_colors(species_names, speciesColors, greys = F, "Pastel1")
                       ), border = T
                    )
community_names <- colnames(abundance_mat_alq) %>% as.data.frame() %>% left_join(psmelt(ps_raw) %>% ungroup() %>% select(OTU, Community) %>% unique, by = c(. = "OTU")) %>% pull(Community)
community_inferred_names <- colnames(abundance_mat_alq) %>% as.data.frame() %>% left_join(adjusted_copy_numbers_aliquots %>% ungroup() %>% select(OTU, Community_inferred) %>% unique, by = c(. = "OTU")) %>% pull(Community_inferred)
column_ha = HeatmapAnnotation(community = community_names,
                              community_inferred = community_inferred_names,
                              col = list(community = communityColors,
                                          community_inferred = communityColors), border = T
                    )
column_split <- community_inferred_names %>% as.factor
row_split <- treatment_names %>% as.factor
range_seq <- c(0, 100, 1000, 10000, 100000)
heatmap_obj = Heatmap(abundance_mat_alq,
        col = colorRamp2(range_seq,
                         colors = c(brewer.pal(9, "GnBu")[1],
                                    brewer.pal(9, "GnBu")[4],
                                    brewer.pal(9, "GnBu")[6],
                                    brewer.pal(9, "GnBu")[8],
                                    brewer.pal(9, "GnBu")[9]
                                    )
                                  ),
        column_split = column_split,
        row_split = row_split,
        row_title_rot = 0,
        column_title_rot = 45,
        # column_title_gp = grid::gpar(fontsize = 0),
        # row_title_gp = grid::gpar(fontsize = 0),
        border = T,
        heatmap_legend_param = list(title = "Absolute Abundance"),
        top_annotation = column_ha, 
        # right_annotation = row_ha,
        bottom_annotation = column_ba,
        column_names_gp = grid::gpar(fontsize = 0),
        row_names_gp = grid::gpar(fontsize = 10),
        cluster_rows = F,
        cluster_columns = F
        )
pdf("04_Results/Figures/Heatmap_ASVs_aliquots_absolute_abundance.pdf", width = 11, height = 8)
draw(heatmap_obj, merge_legend = F, heatmap_legend_side = "right", annotation_legend_side = "left")
dev.off()
draw(heatmap_obj, merge_legend = F, heatmap_legend_side = "right", annotation_legend_side = "left")

column_split <- species_names %>% as.factor
row_split <- treatment_names %>% as.factor
range_seq <- c(0, 100, 1000, 10000, 100000)
heatmap_obj = Heatmap(abundance_mat_alq,
        col = colorRamp2(range_seq,
                         colors = c(brewer.pal(9, "GnBu")[1],
                                    brewer.pal(9, "GnBu")[4],
                                    brewer.pal(9, "GnBu")[6],
                                    brewer.pal(9, "GnBu")[8],
                                    brewer.pal(9, "GnBu")[9]
                                    )
                                  ),
        column_split = column_split,
        row_split = row_split,
        row_title_rot = 0,
        column_title_rot = 65,
        column_title_gp = grid::gpar(fontsize = 10),
        # row_title_gp = grid::gpar(fontsize = 0),
        border = T,
        heatmap_legend_param = list(title = "Absolute Abundance"),
        top_annotation = column_ba, 
        # right_annotation = row_ha,
        bottom_annotation = column_ha,
        column_names_gp = grid::gpar(fontsize = 0),
        row_names_gp = grid::gpar(fontsize = 10, position = "bottom"),
        cluster_rows = F,
        cluster_columns = F
        )
pdf("04_Results/Figures/Heatmap_ASVs_aliquots_absolute_abundance_by_species.pdf", width = 11, height = 8)
draw(heatmap_obj, merge_legend = F, heatmap_legend_side = "right", annotation_legend_side = "left")
dev.off()
draw(heatmap_obj, merge_legend = F, heatmap_legend_side = "right", annotation_legend_side = "left")

column_split <- SDP_names %>% as.factor
row_split <- treatment_names %>% as.factor
range_seq <- c(0, 100, 1000, 10000, 100000)
heatmap_obj = Heatmap(abundance_mat_alq,
        col = colorRamp2(range_seq,
                         colors = c(brewer.pal(9, "GnBu")[1],
                                    brewer.pal(9, "GnBu")[4],
                                    brewer.pal(9, "GnBu")[6],
                                    brewer.pal(9, "GnBu")[8],
                                    brewer.pal(9, "GnBu")[9]
                                    )
                                  ),
        column_split = column_split,
        row_split = row_split,
        row_title_rot = 0,
        column_title_rot = 45,
        column_title_gp = grid::gpar(fontsize = 8),
        # row_title_gp = grid::gpar(fontsize = 0),
        border = T,
        heatmap_legend_param = list(title = "Absolute Abundance"),
        top_annotation = column_ba, 
        # right_annotation = row_ha,
        bottom_annotation = column_ha,
        column_names_gp = grid::gpar(fontsize = 0),
        row_names_gp = grid::gpar(fontsize = 10, position = "bottom"),
        cluster_rows = F,
        cluster_columns = F
        )
pdf("04_Results/Figures/Heatmap_ASVs_aliquots_absolute_abundance_by_SDP.pdf", width = 11, height = 8)
draw(heatmap_obj, merge_legend = F, heatmap_legend_side = "right", annotation_legend_side = "left")
dev.off()
draw(heatmap_obj, merge_legend = F, heatmap_legend_side = "right", annotation_legend_side = "left")

column_split <- strain_name %>% as.factor
row_split <- treatment_names %>% as.factor
range_seq <- c(0, 100, 1000, 10000, 100000)
heatmap_obj = Heatmap(abundance_mat_alq,
        col = colorRamp2(range_seq,
                         colors = c(brewer.pal(9, "GnBu")[1],
                                    brewer.pal(9, "GnBu")[4],
                                    brewer.pal(9, "GnBu")[6],
                                    brewer.pal(9, "GnBu")[8],
                                    brewer.pal(9, "GnBu")[9]
                                    )
                                  ),
        column_split = column_split,
        row_split = row_split,
        row_title_rot = 0,
        column_title_rot = 65,
        column_title_gp = grid::gpar(fontsize = 8),
        # row_title_gp = grid::gpar(fontsize = 0),
        border = T,
        heatmap_legend_param = list(title = "Absolute Abundance"),
        top_annotation = column_ba, 
        # right_annotation = row_ha,
        bottom_annotation = column_ha,
        column_names_gp = grid::gpar(fontsize = 0),
        row_names_gp = grid::gpar(fontsize = 10),
        cluster_rows = F,
        cluster_columns = F
        )
pdf("04_Results/Figures/Heatmap_ASVs_aliquots_absolute_abundance_by_strain.pdf", width = 11, height = 8)
draw(heatmap_obj, merge_legend = F, heatmap_legend_side = "right", annotation_legend_side = "left")
dev.off()
draw(heatmap_obj, merge_legend = F, heatmap_legend_side = "right", annotation_legend_side = "left")

adjusted_copy_numbers_samples <- adjusted_copy_numbers %>%
                                    filter(!grepl("alq", Treatment)) %>%
                                    unique() %>%
                                      left_join(adjusted_copy_numbers_aliquots %>% select(OTU, Community_inferred), by = "OTU")

abundance_mat_samples <- adjusted_copy_numbers_samples %>%
                      select(Sample, OTU, abs_ab) %>%
                        unique %>%
                        pivot_wider(names_from = OTU, values_from = abs_ab, values_fn = as.numeric) %>%
                        replace(is.na(.), 0) %>%
                        column_to_rownames("Sample") %>%
                        as.matrix                             
treatment_names <- rownames(abundance_mat_samples) %>% as.data.frame() %>% left_join(psmelt(ps) %>% ungroup() %>% select(Sample, Treatment) %>% unique, by = c(. = "Sample")) %>% pull(Treatment)
genus_names <- colnames(abundance_mat_samples) %>% as.data.frame() %>% left_join(psmelt(ps) %>% ungroup() %>% select(OTU, Genus) %>% unique, by = c(. = "OTU")) %>% pull(Genus)
species_names <- colnames(abundance_mat_samples) %>% as.data.frame() %>% left_join(psmelt(ps) %>% ungroup() %>% select(OTU, Species, Genus) %>% unique %>% mutate(Species_name = paste0(Genus, " ", Species)), by = c(. = "OTU")) %>% pull(Species_name)
SDP_names <- colnames(abundance_mat_samples) %>% as.data.frame() %>% left_join(psmelt(ps) %>% ungroup() %>% select(OTU, SDP) %>% unique, by = c(. = "OTU")) %>% pull(SDP)
strain_names <- colnames(abundance_mat_samples) %>% as.data.frame() %>% left_join(psmelt(ps) %>% ungroup() %>% select(OTU, SDP, Strain) %>% unique %>% mutate(Strain_name = paste0(SDP, "--", Strain)), by = c(. = "OTU")) %>% pull(Strain_name)
column_ba = HeatmapAnnotation(SDP = SDP_names,
                              Species = species_names,
                              # Genus = genus_names,
                       col = list(SDP = SDPColors,
                                  Species = extend_colors(species_names, speciesColors, greys = F, "Pastel1")
                                  # Genus = genusColors
                       ), border = T
                    )
community_names <- colnames(abundance_mat_samples) %>% as.data.frame() %>% left_join(psmelt(ps_raw) %>% ungroup() %>% select(OTU, Community) %>% unique, by = c(. = "OTU")) %>% pull(Community)
community_inferred_names <- colnames(abundance_mat_samples) %>% as.data.frame() %>% left_join(adjusted_copy_numbers_aliquots %>% ungroup() %>% select(OTU, Community_inferred) %>% unique, by = c(. = "OTU")) %>% pull(Community_inferred)
column_ha = HeatmapAnnotation(community = community_names,
                              community_inferred = community_inferred_names,
                              col = list(community = communityColors,
                                          community_inferred = communityColors), border = T
                    )
column_split <- community_inferred_names %>% as.factor
row_split <- treatment_names %>% as.factor
range_seq <- c(0, 1000, 10000, 100000, 1000000, 1000000, 10000000)
heatmap_obj = Heatmap(abundance_mat_samples,
        col = colorRamp2(range_seq, 
                         colors = c(brewer.pal(9, "GnBu")[1],
                                    brewer.pal(9, "GnBu")[2],
                                    brewer.pal(9, "GnBu")[3],
                                    brewer.pal(9, "GnBu")[4],
                                    brewer.pal(9, "GnBu")[7],
                                    brewer.pal(9, "GnBu")[8],
                                    brewer.pal(9, "GnBu")[9]
                                    )
                                  ),
        column_split = column_split,
        row_split = row_split,
        row_title_rot = 0,
        column_title_rot = 45,
        # column_title_gp = grid::gpar(fontsize = 0),
        # row_title_gp = grid::gpar(fontsize = 0),
        border = T,
        heatmap_legend_param = list(title = "Absolute Abundance"),
        top_annotation = column_ha, 
        # right_annotation = row_ha,
        bottom_annotation = column_ba,
        column_names_gp = grid::gpar(fontsize = 0),
        row_names_gp = grid::gpar(fontsize = 10),
        cluster_rows = F,
        cluster_columns = F
        )
pdf("04_Results/Figures/Heatmap_ASVs_absolute_abundance.pdf", width = 11, height = 8)
draw(heatmap_obj, merge_legend = F, heatmap_legend_side = "right", annotation_legend_side = "left")
dev.off()
draw(heatmap_obj, merge_legend = F, heatmap_legend_side = "right", annotation_legend_side = "left")

column_split <- strain_names %>% as.factor
row_split <- treatment_names %>% as.factor
range_seq <- c(0, 1000, 10000, 100000, 1000000, 1000000, 10000000)
heatmap_obj = Heatmap(abundance_mat_samples,
        col = colorRamp2(range_seq, 
                         colors = c(brewer.pal(9, "GnBu")[1],
                                    brewer.pal(9, "GnBu")[2],
                                    brewer.pal(9, "GnBu")[3],
                                    brewer.pal(9, "GnBu")[4],
                                    brewer.pal(9, "GnBu")[7],
                                    brewer.pal(9, "GnBu")[8],
                                    brewer.pal(9, "GnBu")[9]
                                    )
                                  ),
        column_split = column_split,
        row_split = row_split,
        row_title_rot = 0,
        column_title_rot = 65,
        column_title_gp = grid::gpar(fontsize = 6),
        # row_title_gp = grid::gpar(fontsize = 0),
        border = T,
        heatmap_legend_param = list(title = "Absolute Abundance"),
        top_annotation = column_ba, 
        # right_annotation = row_ha,
        bottom_annotation = column_ha,
        column_names_gp = grid::gpar(fontsize = 0),
        row_names_gp = grid::gpar(fontsize = 10),
        cluster_rows = F,
        cluster_columns = F
        )
pdf("04_Results/Figures/Heatmap_ASVs_absolute_abundance_by_strain.pdf", width = 11, height = 8)
draw(heatmap_obj, merge_legend = F, heatmap_legend_side = "right", annotation_legend_side = "left")
dev.off()
draw(heatmap_obj, merge_legend = F, heatmap_legend_side = "right", annotation_legend_side = "left")

column_split <- SDP_names %>% as.factor
row_split <- treatment_names %>% as.factor
range_seq <- c(0, 1000, 10000, 100000, 1000000, 1000000, 10000000)
heatmap_obj = Heatmap(abundance_mat_samples,
        col = colorRamp2(range_seq, 
                         colors = c(brewer.pal(9, "GnBu")[1],
                                    brewer.pal(9, "GnBu")[2],
                                    brewer.pal(9, "GnBu")[3],
                                    brewer.pal(9, "GnBu")[4],
                                    brewer.pal(9, "GnBu")[7],
                                    brewer.pal(9, "GnBu")[8],
                                    brewer.pal(9, "GnBu")[9]
                                    )
                                  ),
        column_split = column_split,
        row_split = row_split,
        row_title_rot = 0,
        column_title_rot = 45,
        column_title_gp = grid::gpar(fontsize = 6),
        # row_title_gp = grid::gpar(fontsize = 0),
        border = T,
        heatmap_legend_param = list(title = "Absolute Abundance"),
        top_annotation = column_ba, 
        # right_annotation = row_ha,
        bottom_annotation = column_ha,
        column_names_gp = grid::gpar(fontsize = 0),
        row_names_gp = grid::gpar(fontsize = 10),
        cluster_rows = F,
        cluster_columns = F
        )
pdf("04_Results/Figures/Heatmap_ASVs_absolute_abundance_by_SDP.pdf", width = 11, height = 8)
draw(heatmap_obj, merge_legend = F, heatmap_legend_side = "right", annotation_legend_side = "left")
dev.off()
draw(heatmap_obj, merge_legend = F, heatmap_legend_side = "right", annotation_legend_side = "left")

column_split <- genus_names %>% as.factor
row_split <- treatment_names %>% as.factor
range_seq <- c(0, 1000, 10000, 100000, 1000000, 1000000, 10000000)
heatmap_obj = Heatmap(abundance_mat_samples,
        col = colorRamp2(range_seq, 
                         colors = c(brewer.pal(9, "GnBu")[1],
                                    brewer.pal(9, "GnBu")[2],
                                    brewer.pal(9, "GnBu")[3],
                                    brewer.pal(9, "GnBu")[4],
                                    brewer.pal(9, "GnBu")[7],
                                    brewer.pal(9, "GnBu")[8],
                                    brewer.pal(9, "GnBu")[9]
                                    )
                                  ),
        column_split = column_split,
        row_split = row_split,
        row_title_rot = 0,
        column_title_rot = 65,
        column_title_gp = grid::gpar(fontsize = 8),
        # row_title_gp = grid::gpar(fontsize = 0),
        border = T,
        heatmap_legend_param = list(title = "Absolute Abundance"),
        top_annotation = column_ba, 
        # right_annotation = row_ha,
        bottom_annotation = column_ha,
        column_names_gp = grid::gpar(fontsize = 0),
        row_names_gp = grid::gpar(fontsize = 10),
        cluster_rows = F,
        cluster_columns = F
        )
pdf("04_Results/Figures/Heatmap_ASVs_absolute_abundance_by_genus.pdf", width = 11, height = 8)
draw(heatmap_obj, merge_legend = F, heatmap_legend_side = "right", annotation_legend_side = "left")
dev.off()
draw(heatmap_obj, merge_legend = F, heatmap_legend_side = "right", annotation_legend_side = "left")
```
```{r}
df_plot_samples <- adjusted_copy_numbers_samples %>%
  select(OTU, Sample, Treatment, Community_inferred, abs_ab, rel_ab, Genus) %>%
    unique() %>%
    mutate(Community_final = ifelse(Community_inferred == "B_ambiguous", "ambiguous", Community_inferred)) %>%
    mutate(Community_final = ifelse(Community_inferred == "AB_ambiguous", "ambiguous", Community_final)) %>%
    mutate(Community_final = ifelse(Community_inferred == "A_ambiguous", "ambiguous", Community_final)) %>%
    mutate(Community_final = ifelse(Community_inferred == "unknown", Genus, Community_final)) %>%
      group_by(OTU, Treatment) %>%
        mutate(abs_ab_treatment = mean(abs_ab)) %>%
        ungroup() %>%
        group_by(Treatment) %>%
        mutate(rel_ab_treatment = abs_ab_treatment/sum(abs_ab_treatment)*100)

ggplot() +
  geom_bar(data = df_plot_samples,
           aes(x = factor(Sample, ID_order),
               y = rel_ab,
               fill = Community_final
              #  fill = factor(Community_final, c("A", "B", "AB"))
          ), stat = "identity", position = "stack", color = "black", size = 0.1
  ) +
  labs(x = "Sample", y = "Absolute abundance", fill = "Community    ") +
  scale_fill_manual(values = unlist(extend_colors(df_plot_samples$Community_final, communityColors_light, pal = "Spectral"))) +
  facet_wrap(~Treatment, scale = "free_x") +
    make_theme(setCol = F, setFill = F, palettecolor = "Pastel1",
               axis_x_title = 18,
               axis_y_title = 18, guide_nrow = 2,
               y_size = 15,
               x_size = 10)
    ggsave("04_Results/Figures/per_sample_rel_ab_by_treatment.pdf")

ggplot() +
  geom_bar(data = df_plot_samples,
           aes(x = factor(Sample, ID_order),
               y = rel_ab,
               fill = Community_inferred
          ), stat = "identity", position = "stack", color = "black", size = 0.1
  ) +
  labs(x = "Sample", y = "Absolute abundance", fill = "Community    ") +
  scale_fill_manual(values = unlist(extend_colors(df_plot_samples$Community_inferred, communityColors_light, pal = "Spectral"))) +
  facet_wrap(~Treatment, scale = "free_x") +
    make_theme(setCol = F, setFill = F, palettecolor = "Pastel1",
               axis_x_title = 18,
               axis_y_title = 18, guide_nrow = 2,
               y_size = 15,
               x_size = 10)
    ggsave("04_Results/Figures/per_sample_rel_ab_by_treatment_ambiguous_included.pdf")

ggplot() +
  geom_bar(data = df_plot_samples,
           aes(x = factor(Sample, ID_order),
               y = abs_ab,
               fill = Community_inferred
              #  fill = factor(Community_final, c("A", "B", "AB"))
          ), stat = "identity", position = "stack", color = "black", size = 0.1
  ) +
  labs(x = "Sample", y = "Absolute abundance", fill = "Community    ") +
  scale_fill_manual(values = unlist(extend_colors(df_plot_samples$Community_inferred, communityColors_light, pal = "Spectral"))) +
  facet_wrap(~Treatment, scale = "free_x") +
    # scale_y_continuous(labels = function(x) parse(text=paste("10^",x))) +
    make_theme(setCol = F, setFill = F, palettecolor = "Pastel1",
               axis_x_title = 18,
               axis_y_title = 18, guide_nrow = 1,
               y_size = 15,
               x_size = 10)
    ggsave("04_Results/Figures/per_sample_abs_ab_by_treatment.pdf")

ggplot() +
  geom_bar(data = df_plot_samples,
           aes(x = factor(Treatment, treatment_order),
               y = abs_ab_treatment,
               fill = Community_inferred
          ), stat = "identity", position = "stack", color = "black", size = 0.1
  ) +
  labs(x = "Sample", y = "Absolute abundance", fill = "Community    ") +
  scale_fill_manual(values = unlist(extend_colors(df_plot_samples$Community_inferred, communityColors_light, greys = F, pal = "Pastel1"))) +
    # scale_y_continuous(labels = function(x) parse(text=paste("10^",x))) +
    make_theme(setCol = F, setFill = F,
               axis_x_title = 18,
               axis_y_title = 18, guide_nrow = 3,
               y_size = 15,
               x_size = 10)
    ggsave("04_Results/Figures/mean_abs_ab_by_treatment.pdf")

ggplot() +
  geom_bar(data = df_plot_samples,
           aes(x = factor(Treatment, treatment_order),
               y = rel_ab_treatment,
               fill = Community_inferred
          ), stat = "identity", position = "stack", color = "black", size = 0.1
  ) +
  labs(x = "Sample", y = "Absolute abundance", fill = "Community    ") +
  scale_fill_manual(values = unlist(extend_colors(df_plot_samples$Community_inferred, communityColors_light, greys = F, pal = "Pastel1"))) +
    # scale_y_continuous(labels = function(x) parse(text=paste("10^",x))) +
    make_theme(setCol = F, setFill = F,
               axis_x_title = 18,
               axis_y_title = 18, guide_nrow = 3,
               y_size = 15,
               x_size = 10)
    ggsave("04_Results/Figures/mean_rel_ab_by_treatment.pdf")
```