cd /home/aiswarya/mnt/nas_recherche/spirit/D2c/aprasad/20220921_aprasad_PriorityEffectsExperimentPilot/03_PilotExperiment/07_PacBIO_sequencing
conda activate pacbio-ampli-env

00_RawData/AP_16S_Circular_Consensus_Sequencing_Reads

 cp -n 00_RawData/AP_16S_Circular_Consensus_Sequencing_Reads/*.bc1005--bc1033.hifi_reads.fastq.gz  01_ReadsRenamed/1-1_reads.fastq.gz
 cp -n 00_RawData/AP_16S_Circular_Consensus_Sequencing_Reads/*.bc1007--bc1033.hifi_reads.fastq.gz  01_ReadsRenamed/1-2_reads.fastq.gz
 cp -n 00_RawData/AP_16S_Circular_Consensus_Sequencing_Reads/*.bc1008--bc1033.hifi_reads.fastq.gz  01_ReadsRenamed/1-3_reads.fastq.gz
 cp -n 00_RawData/AP_16S_Circular_Consensus_Sequencing_Reads/*.bc1012--bc1033.hifi_reads.fastq.gz  01_ReadsRenamed/1-4_reads.fastq.gz
 cp -n 00_RawData/AP_16S_Circular_Consensus_Sequencing_Reads/*.bc1015--bc1033.hifi_reads.fastq.gz  01_ReadsRenamed/1-5_reads.fastq.gz
 cp -n 00_RawData/AP_16S_Circular_Consensus_Sequencing_Reads/*.bc1020--bc1033.hifi_reads.fastq.gz  01_ReadsRenamed/1-6_reads.fastq.gz
 cp -n 00_RawData/AP_16S_Circular_Consensus_Sequencing_Reads/*.bc1022--bc1033.hifi_reads.fastq.gz  01_ReadsRenamed/1-7_reads.fastq.gz
 cp -n 00_RawData/AP_16S_Circular_Consensus_Sequencing_Reads/*.bc1024--bc1033.hifi_reads.fastq.gz  01_ReadsRenamed/1-8_reads.fastq.gz
 cp -n 00_RawData/AP_16S_Circular_Consensus_Sequencing_Reads/*.bc1005--bc1035.hifi_reads.fastq.gz  01_ReadsRenamed/1-9_reads.fastq.gz
 cp -n 00_RawData/AP_16S_Circular_Consensus_Sequencing_Reads/*.bc1007--bc1035.hifi_reads.fastq.gz  01_ReadsRenamed/1-10_reads.fastq.gz
 cp -n 00_RawData/AP_16S_Circular_Consensus_Sequencing_Reads/*.bc1008--bc1035.hifi_reads.fastq.gz  01_ReadsRenamed/1-11_reads.fastq.gz
 cp -n 00_RawData/AP_16S_Circular_Consensus_Sequencing_Reads/*.bc1012--bc1035.hifi_reads.fastq.gz  01_ReadsRenamed/1-12_reads.fastq.gz
 cp -n 00_RawData/AP_16S_Circular_Consensus_Sequencing_Reads/*.bc1015--bc1035.hifi_reads.fastq.gz  01_ReadsRenamed/1-13_reads.fastq.gz
 cp -n 00_RawData/AP_16S_Circular_Consensus_Sequencing_Reads/*.bc1020--bc1035.hifi_reads.fastq.gz  01_ReadsRenamed/1-14_reads.fastq.gz
 cp -n 00_RawData/AP_16S_Circular_Consensus_Sequencing_Reads/*.bc1022--bc1035.hifi_reads.fastq.gz  01_ReadsRenamed/1-15_reads.fastq.gz
 cp -n 00_RawData/AP_16S_Circular_Consensus_Sequencing_Reads/*.bc1024--bc1035.hifi_reads.fastq.gz  01_ReadsRenamed/1-16_reads.fastq.gz
 cp -n 00_RawData/AP_16S_Circular_Consensus_Sequencing_Reads/*.bc1005--bc1044.hifi_reads.fastq.gz  01_ReadsRenamed/1-17_reads.fastq.gz
 cp -n 00_RawData/AP_16S_Circular_Consensus_Sequencing_Reads/*.bc1007--bc1044.hifi_reads.fastq.gz  01_ReadsRenamed/1-18_reads.fastq.gz
 cp -n 00_RawData/AP_16S_Circular_Consensus_Sequencing_Reads/*.bc1008--bc1044.hifi_reads.fastq.gz  01_ReadsRenamed/1-19_reads.fastq.gz
 cp -n 00_RawData/AP_16S_Circular_Consensus_Sequencing_Reads/*.bc1012--bc1044.hifi_reads.fastq.gz  01_ReadsRenamed/1-20_reads.fastq.gz
 cp -n 00_RawData/AP_16S_Circular_Consensus_Sequencing_Reads/*.bc1015--bc1044.hifi_reads.fastq.gz  01_ReadsRenamed/1-21_reads.fastq.gz
 cp -n 00_RawData/AP_16S_Circular_Consensus_Sequencing_Reads/*.bc1020--bc1044.hifi_reads.fastq.gz  01_ReadsRenamed/1-22_reads.fastq.gz
 cp -n 00_RawData/AP_16S_Circular_Consensus_Sequencing_Reads/*.bc1022--bc1044.hifi_reads.fastq.gz  01_ReadsRenamed/1-23_reads.fastq.gz
 cp -n 00_RawData/AP_16S_Circular_Consensus_Sequencing_Reads/*.bc1024--bc1044.hifi_reads.fastq.gz  01_ReadsRenamed/1-24_reads.fastq.gz
 cp -n 00_RawData/AP_16S_Circular_Consensus_Sequencing_Reads/*.bc1005--bc1045.hifi_reads.fastq.gz  01_ReadsRenamed/2-1_reads.fastq.gz
 cp -n 00_RawData/AP_16S_Circular_Consensus_Sequencing_Reads/*.bc1007--bc1045.hifi_reads.fastq.gz  01_ReadsRenamed/2-2_reads.fastq.gz
 cp -n 00_RawData/AP_16S_Circular_Consensus_Sequencing_Reads/*.bc1008--bc1045.hifi_reads.fastq.gz  01_ReadsRenamed/2-3_reads.fastq.gz
 cp -n 00_RawData/AP_16S_Circular_Consensus_Sequencing_Reads/*.bc1012--bc1045.hifi_reads.fastq.gz  01_ReadsRenamed/2-4_reads.fastq.gz
 cp -n 00_RawData/AP_16S_Circular_Consensus_Sequencing_Reads/*.bc1015--bc1045.hifi_reads.fastq.gz  01_ReadsRenamed/2-5_reads.fastq.gz
 cp -n 00_RawData/AP_16S_Circular_Consensus_Sequencing_Reads/*.bc1020--bc1045.hifi_reads.fastq.gz  01_ReadsRenamed/2-6_reads.fastq.gz
 cp -n 00_RawData/AP_16S_Circular_Consensus_Sequencing_Reads/*.bc1022--bc1045.hifi_reads.fastq.gz  01_ReadsRenamed/2-7_reads.fastq.gz
 cp -n 00_RawData/AP_16S_Circular_Consensus_Sequencing_Reads/*.bc1024--bc1045.hifi_reads.fastq.gz  01_ReadsRenamed/2-8_reads.fastq.gz
 cp -n 00_RawData/AP_16S_Circular_Consensus_Sequencing_Reads/*.bc1005--bc1054.hifi_reads.fastq.gz  01_ReadsRenamed/2-9_reads.fastq.gz
 cp -n 00_RawData/AP_16S_Circular_Consensus_Sequencing_Reads/*.bc1007--bc1054.hifi_reads.fastq.gz  01_ReadsRenamed/2-10_reads.fastq.gz
 cp -n 00_RawData/AP_16S_Circular_Consensus_Sequencing_Reads/*.bc1008--bc1054.hifi_reads.fastq.gz  01_ReadsRenamed/2-11_reads.fastq.gz
 cp -n 00_RawData/AP_16S_Circular_Consensus_Sequencing_Reads/*.bc1012--bc1054.hifi_reads.fastq.gz  01_ReadsRenamed/2-12_reads.fastq.gz
 cp -n 00_RawData/AP_16S_Circular_Consensus_Sequencing_Reads/*.bc1015--bc1054.hifi_reads.fastq.gz  01_ReadsRenamed/2-13_reads.fastq.gz
 cp -n 00_RawData/AP_16S_Circular_Consensus_Sequencing_Reads/*.bc1020--bc1054.hifi_reads.fastq.gz  01_ReadsRenamed/2-14_reads.fastq.gz
 cp -n 00_RawData/AP_16S_Circular_Consensus_Sequencing_Reads/*.bc1022--bc1054.hifi_reads.fastq.gz  01_ReadsRenamed/2-15_reads.fastq.gz
 cp -n 00_RawData/AP_16S_Circular_Consensus_Sequencing_Reads/*.bc1024--bc1054.hifi_reads.fastq.gz  01_ReadsRenamed/2-16_reads.fastq.gz
 cp -n 00_RawData/AP_16S_Circular_Consensus_Sequencing_Reads/*.bc1005--bc1056.hifi_reads.fastq.gz  01_ReadsRenamed/2-17_reads.fastq.gz
 cp -n 00_RawData/AP_16S_Circular_Consensus_Sequencing_Reads/*.bc1007--bc1056.hifi_reads.fastq.gz  01_ReadsRenamed/2-18_reads.fastq.gz
 cp -n 00_RawData/AP_16S_Circular_Consensus_Sequencing_Reads/*.bc1008--bc1056.hifi_reads.fastq.gz  01_ReadsRenamed/2-19_reads.fastq.gz
 cp -n 00_RawData/AP_16S_Circular_Consensus_Sequencing_Reads/*.bc1012--bc1056.hifi_reads.fastq.gz  01_ReadsRenamed/2-20_reads.fastq.gz
 cp -n 00_RawData/AP_16S_Circular_Consensus_Sequencing_Reads/*.bc1015--bc1056.hifi_reads.fastq.gz  01_ReadsRenamed/2-21_reads.fastq.gz
 cp -n 00_RawData/AP_16S_Circular_Consensus_Sequencing_Reads/*.bc1020--bc1056.hifi_reads.fastq.gz  01_ReadsRenamed/2-22_reads.fastq.gz
 cp -n 00_RawData/AP_16S_Circular_Consensus_Sequencing_Reads/*.bc1022--bc1056.hifi_reads.fastq.gz  01_ReadsRenamed/2-23_reads.fastq.gz
 cp -n 00_RawData/AP_16S_Circular_Consensus_Sequencing_Reads/*.bc1024--bc1056.hifi_reads.fastq.gz  01_ReadsRenamed/2-24_reads.fastq.gz
 cp -n 00_RawData/AP_16S_Circular_Consensus_Sequencing_Reads/*.bc1005--bc1057.hifi_reads.fastq.gz  01_ReadsRenamed/3-1_reads.fastq.gz
 cp -n 00_RawData/AP_16S_Circular_Consensus_Sequencing_Reads/*.bc1007--bc1057.hifi_reads.fastq.gz  01_ReadsRenamed/3-2_reads.fastq.gz
 cp -n 00_RawData/AP_16S_Circular_Consensus_Sequencing_Reads/*.bc1008--bc1057.hifi_reads.fastq.gz  01_ReadsRenamed/3-3_reads.fastq.gz
 cp -n 00_RawData/AP_16S_Circular_Consensus_Sequencing_Reads/*.bc1012--bc1057.hifi_reads.fastq.gz  01_ReadsRenamed/3-4_reads.fastq.gz
 cp -n 00_RawData/AP_16S_Circular_Consensus_Sequencing_Reads/*.bc1015--bc1057.hifi_reads.fastq.gz  01_ReadsRenamed/3-5_reads.fastq.gz
 cp -n 00_RawData/AP_16S_Circular_Consensus_Sequencing_Reads/*.bc1020--bc1057.hifi_reads.fastq.gz  01_ReadsRenamed/3-6_reads.fastq.gz
 cp -n 00_RawData/AP_16S_Circular_Consensus_Sequencing_Reads/*.bc1022--bc1057.hifi_reads.fastq.gz  01_ReadsRenamed/3-7_reads.fastq.gz
 cp -n 00_RawData/AP_16S_Circular_Consensus_Sequencing_Reads/*.bc1024--bc1057.hifi_reads.fastq.gz  01_ReadsRenamed/3-8_reads.fastq.gz
 cp -n 00_RawData/AP_16S_Circular_Consensus_Sequencing_Reads/*.bc1005--bc1059.hifi_reads.fastq.gz  01_ReadsRenamed/3-9_reads.fastq.gz
 cp -n 00_RawData/AP_16S_Circular_Consensus_Sequencing_Reads/*.bc1007--bc1059.hifi_reads.fastq.gz  01_ReadsRenamed/3-10_reads.fastq.gz
 cp -n 00_RawData/AP_16S_Circular_Consensus_Sequencing_Reads/*.bc1008--bc1059.hifi_reads.fastq.gz  01_ReadsRenamed/3-11_reads.fastq.gz
 cp -n 00_RawData/AP_16S_Circular_Consensus_Sequencing_Reads/*.bc1012--bc1059.hifi_reads.fastq.gz  01_ReadsRenamed/3-12_reads.fastq.gz
 cp -n 00_RawData/AP_16S_Circular_Consensus_Sequencing_Reads/*.bc1015--bc1059.hifi_reads.fastq.gz  01_ReadsRenamed/3-13_reads.fastq.gz
 cp -n 00_RawData/AP_16S_Circular_Consensus_Sequencing_Reads/*.bc1020--bc1059.hifi_reads.fastq.gz  01_ReadsRenamed/3-14_reads.fastq.gz
 cp -n 00_RawData/AP_16S_Circular_Consensus_Sequencing_Reads/*.bc1022--bc1059.hifi_reads.fastq.gz  01_ReadsRenamed/3-15_reads.fastq.gz
 cp -n 00_RawData/AP_16S_Circular_Consensus_Sequencing_Reads/*.bc1024--bc1059.hifi_reads.fastq.gz  01_ReadsRenamed/3-16_reads.fastq.gz
 cp -n 00_RawData/AP_16S_Circular_Consensus_Sequencing_Reads/*.bc1005--bc1060.hifi_reads.fastq.gz  01_ReadsRenamed/3-17_reads.fastq.gz
 cp -n 00_RawData/AP_16S_Circular_Consensus_Sequencing_Reads/*.bc1007--bc1060.hifi_reads.fastq.gz  01_ReadsRenamed/3-18_reads.fastq.gz
 cp -n 00_RawData/AP_16S_Circular_Consensus_Sequencing_Reads/*.bc1008--bc1060.hifi_reads.fastq.gz  01_ReadsRenamed/3-19_reads.fastq.gz
 cp -n 00_RawData/AP_16S_Circular_Consensus_Sequencing_Reads/*.bc1012--bc1060.hifi_reads.fastq.gz  01_ReadsRenamed/3-20_reads.fastq.gz
 cp -n 00_RawData/AP_16S_Circular_Consensus_Sequencing_Reads/*.bc1015--bc1060.hifi_reads.fastq.gz  01_ReadsRenamed/3-21_reads.fastq.gz
 cp -n 00_RawData/AP_16S_Circular_Consensus_Sequencing_Reads/*.bc1020--bc1060.hifi_reads.fastq.gz  01_ReadsRenamed/3-22_reads.fastq.gz
 cp -n 00_RawData/AP_16S_Circular_Consensus_Sequencing_Reads/*.bc1022--bc1060.hifi_reads.fastq.gz  01_ReadsRenamed/LW-1_reads.fastq.gz
 cp -n 00_RawData/AP_16S_Circular_Consensus_Sequencing_Reads/*.bc1024--bc1060.hifi_reads.fastq.gz  01_ReadsRenamed/NC2-1_reads.fastq.gz
 cp -n 00_RawData/AP_16S_Circular_Consensus_Sequencing_Reads/*.bc1005--bc1062.hifi_reads.fastq.gz  01_ReadsRenamed/SC_reads.fastq.gz
 cp -n 00_RawData/AP_16S_Circular_Consensus_Sequencing_Reads/*.bc1007--bc1062.hifi_reads.fastq.gz  01_ReadsRenamed/NC2SC_reads.fastq.gz
 cp -n 00_RawData/AP_16S_Circular_Consensus_Sequencing_Reads/*.bc1008--bc1062.hifi_reads.fastq.gz  01_ReadsRenamed/M6-5_reads.fastq.gz
 cp -n 00_RawData/AP_16S_Circular_Consensus_Sequencing_Reads/*.bc1012--bc1062.hifi_reads.fastq.gz  01_ReadsRenamed/C8-2_reads.fastq.gz
 cp -n 00_RawData/AP_16S_Circular_Consensus_Sequencing_Reads/*.bc1015--bc1062.hifi_reads.fastq.gz  01_ReadsRenamed/D7-4_reads.fastq.gz
 cp -n 00_RawData/AP_16S_Circular_Consensus_Sequencing_Reads/*.bc1020--bc1062.hifi_reads.fastq.gz  01_ReadsRenamed/F4-3_reads.fastq.gz
 cp -n 00_RawData/AP_16S_Circular_Consensus_Sequencing_Reads/*.bc1022--bc1062.hifi_reads.fastq.gz  01_ReadsRenamed/A4-3_reads.fastq.gz

######
# Count reads
######

echo "ID, raw_count" > 01_ReadsRenamed/read_counts.csv
for file in 01_ReadsRenamed/*_reads.fastq.gz;
do
    sample=$(basename ${file})
    sample=${sample%%_reads*}
    echo "counting ${sample}"
    count_raw=$(zcat $file | grep -c "^+$")
    echo "${sample},${count_raw}" >> 01_ReadsRenamed/read_counts.csv
done

mkdir 02_FastQCBeforeTrimming

for file in 01_ReadsRenamed/*_reads.fastq.gz;
do
    sample=$(basename ${file})
    sample=${sample%%_reads*}
    echo "Runnning fastqc for ${sample}"
    fastqc $file -o 02_FastQCBeforeTrimming/
done

# No adpter sequence in fastq reads!

# mkdir 03_CutAdaptReads

# for file in 01_ReadsRenamed/*_reads.fastq.gz;
# do
#     sample=$(basename ${file})
#     sample=${sample%%_reads*}
#     echo "Runnning cutadapt for ${sample}"
# done

# If headers are fine, use the below code to concatenate into multifasta
# The format of the name of headers in the 16S sequences database files is expected to be:
# <strain name>-<# serial number for copy>_<SDP name>.fa
# SDP name may contain "_" and "." but not "-"
# eg. ESL252-1_firm5_1
# for file in  database/16S_sequences/*; do cat $file; echo ""; done | grep -v -e '^$'
# use
# mafft --clustalout --auto --adjustdirection database/16S_sequences/16S_sequences.fasta
# to visualize the output then,
# "_R_" is added if the sequence had to be reversed

python3 scripts/identify_unique_position_set.py --database_path "database/16S_sequences/" \
                                                --input_fasta "database/16S_sequences/16S_sequences.fasta" \
                                                --output_file "database/16S_sequences/unique_position_info.pickle" \
                                                --subset true